R-package

mlampros opened the pull request:
This pull request is the continuation of the #165 issue.

mlampros commented:
@StrikerRUS now that I've added the git sub module any suggestions? I can't extract any info from the #166 on what should be done.

StrikerRUS commented:
So, git created submodule automatically?.. You've just copy-pasted your repo in the new folder, right?

mlampros commented:
yes that's correct I just copy-pasted the repo. Detailed : I did the fork, then clone locally, I copy-pasted my repo to the fork and I uploaded the fork to my account, then I did the pull request.

StrikerRUS commented:
As you can see this resulted in creating the submodule: https://github.com/RGF-team/rgf/pull/208/files

I didn't expect this...

Please try the following: navigate to the `R-package` folder, and remove **hidden** folder `.git`. After that commit the changes.

mlampros commented:
@StrikerRUS any idea how this can be done. I removed the .git folder locally then I pushed the changes however I receive,

```R
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean


```

StrikerRUS commented:
Hmm...

What gives `git status`?

Maybe this can help?

`git rm --cached submodule_path # delete reference to submodule HEAD (no trailing slash)`
`git rm --cached R-package`
from here https://stackoverflow.com/a/1789374

StrikerRUS commented:
I managed to add `R-package` without submoduling:

```
D:\Users\nekit\Downloads>git clone git@github.com:RGF-team/rgf.git
Cloning into 'rgf'...
remote: Counting objects: 1972, done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 1972 (delta 6), reused 5 (delta 1), pack-reused 1951
Receiving objects: 100% (1972/1972), 4.99 MiB | 1.85 MiB/s, done.
Resolving deltas: 100% (1104/1104), done.

D:\Users\nekit\Downloads>git clone git@github.com:mlampros/RGF.git R-RGF
Cloning into 'R-RGF'...
remote: Counting objects: 179, done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 179 (delta 8), reused 19 (delta 6), pack-reused 152
Receiving objects: 100% (179/179), 84.45 KiB | 554.00 KiB/s, done.
Resolving deltas: 100% (46/46), done.

D:\Users\nekit\Downloads>cd rgf

D:\Users\nekit\Downloads\rgf>mkdir R-package

<copy all content of ../R-RGF folder (your repo) into ./R-package folder>

D:\Users\nekit\Downloads\rgf>git status
On branch master
Your branch is up-to-date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        R-package/

nothing added to commit but untracked files present (use "git add" to track)

D:\Users\nekit\Downloads\rgf>git add *

D:\Users\nekit\Downloads\rgf>git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        new file:   R-package/.Rbuildignore
        new file:   R-package/.gitignore
        new file:   R-package/.travis.yml
        new file:   R-package/DESCRIPTION
        new file:   R-package/LICENSE
        new file:   R-package/LICENSE.note
        new file:   R-package/NAMESPACE
        new file:   R-package/NEWS.md
        new file:   R-package/R/package.R
        new file:   R-package/R/utils.R
        new file:   R-package/R/zzz.R
        new file:   R-package/README.md
        new file:   R-package/man/FastRGF_Classifier.Rd
        new file:   R-package/man/FastRGF_Regressor.Rd
        new file:   R-package/man/Internal_class.Rd
        new file:   R-package/man/RGF_Classifier.Rd
        new file:   R-package/man/RGF_Regressor.Rd
        new file:   R-package/man/RGF_cleanup_temp_files.Rd
        new file:   R-package/man/TO_scipy_sparse.Rd
        new file:   R-package/man/mat_2scipy_sparse.Rd
        new file:   R-package/tests/testthat.R
        new file:   R-package/tests/testthat/test_package.R
        new file:   R-package/vignettes/the_RGF_package.Rmd


D:\Users\nekit\Downloads\rgf>git commit -m"added R-package"
[master 51eac09] added R-package
 23 files changed, 3290 insertions(+)
 create mode 100644 R-package/.Rbuildignore
 create mode 100644 R-package/.gitignore
 create mode 100644 R-package/.travis.yml
 create mode 100644 R-package/DESCRIPTION
 create mode 100644 R-package/LICENSE
 create mode 100644 R-package/LICENSE.note
 create mode 100644 R-package/NAMESPACE
 create mode 100644 R-package/NEWS.md
 create mode 100644 R-package/R/package.R
 create mode 100644 R-package/R/utils.R
 create mode 100644 R-package/R/zzz.R
 create mode 100644 R-package/README.md
 create mode 100644 R-package/man/FastRGF_Classifier.Rd
 create mode 100644 R-package/man/FastRGF_Regressor.Rd
 create mode 100644 R-package/man/Internal_class.Rd
 create mode 100644 R-package/man/RGF_Classifier.Rd
 create mode 100644 R-package/man/RGF_Regressor.Rd
 create mode 100644 R-package/man/RGF_cleanup_temp_files.Rd
 create mode 100644 R-package/man/TO_scipy_sparse.Rd
 create mode 100644 R-package/man/mat_2scipy_sparse.Rd
 create mode 100644 R-package/tests/testthat.R
 create mode 100644 R-package/tests/testthat/test_package.R
 create mode 100644 R-package/vignettes/the_RGF_package.Rmd

D:\Users\nekit\Downloads\rgf>git status
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```

StrikerRUS commented:
I've fixed this PR, so we can start review now. :-)

@mlampros 
You can apply my changes locally on your compute by either `git pull`, or just re-clone `rgf-1` repo.

StrikerRUS commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,28 @@
+Date: 2018-07-22
+Authors@R: c( person("Lampros", "Mouselimis", email = "mouselimislampros@gmail.com", role = c("aut", "cre")), person("Ryosuke", "Fukatani", role = "cph", comment = "Author of the python wrapper of the 'Regularized Greedy Forest' machine learning algorithm"), person("Tong", "Zhang", role = "cph", comment = "Author of the 'Regularized Greedy Forest' and of the Multi-core implementation of Regulariz ...
+Maintainer: Lampros Mouselimis <mouselimislampros@gmail.com>
+BugReports: https://github.com/mlampros/RGF/issues
```
Please update the link with `https://github.com/RGF-team/rgf/issues`

mlampros replied:
fixed and pushed changes

StrikerRUS commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,28 @@
+Authors@R: c( person("Lampros", "Mouselimis", email = "mouselimislampros@gmail.com", role = c("aut", "cre")), person("Ryosuke", "Fukatani", role = "cph", comment = "Author of the python wrapper of the 'Regularized Greedy Forest' machine learning algorithm"), person("Tong", "Zhang", role = "cph", comment = "Author of the 'Regularized Greedy Forest' and of the Multi-core implementation of Regulariz ...
+Maintainer: Lampros Mouselimis <mouselimislampros@gmail.com>
+BugReports: https://github.com/mlampros/RGF/issues
+URL: https://github.com/mlampros/RGF
```
`https://github.com/RGF-team/rgf/tree/master/R-package`

mlampros replied:
fixed and pushed changes

StrikerRUS commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,28 @@
+Maintainer: Lampros Mouselimis <mouselimislampros@gmail.com>
+BugReports: https://github.com/mlampros/RGF/issues
+URL: https://github.com/mlampros/RGF
+Description: Regularized Greedy Forest wrapper of the 'Regularized Greedy Forest' <https://github.com/fukatani/rgf_python> 'python' package, which also includes a Multi-core implementation (FastRGF) <https://github.com/baidu/fast_rgf>.
```
Links should be updated

mlampros replied:
fixed and pushed changes

StrikerRUS commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,28 @@
+URL: https://github.com/mlampros/RGF
+Description: Regularized Greedy Forest wrapper of the 'Regularized Greedy Forest' <https://github.com/fukatani/rgf_python> 'python' package, which also includes a Multi-core implementation (FastRGF) <https://github.com/baidu/fast_rgf>.
+License: MIT + file LICENSE
+SystemRequirements: Python (2.7 or >= 3.4), rgf_python, scikit-learn (>= 0.18.0), scipy ( >= 1.0.0), numpy ( >= 1.14.0). Detailed installation instructions for each operating system can be found in the README file.
```
>scipy ( >= 1.0.0), numpy ( >= 1.14.0).

Can you please explain this?

mlampros replied:
The RGF R package includes the *mat_2scipy_sparse* and *TO_scipy_sparse* functions. For these two functions to work properly, scipy and numpy are the requirements.

StrikerRUS replied:
Yeah, sure! rgf_python requires these packages too, because `scikit-learn` requires them =)
I'm wondering about such strange (very fresh) **versions**.

StrikerRUS commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,28 @@
+VignetteBuilder: knitr
+
+
+ 
```
Please remove 3 excess blank lines in the end of the file

mlampros replied:
fixed and pushed changes

mlampros commented:
@StrikerRUS so that I'm sure that I updated the correct DESCRIPTION file. Inside the R-package folder now appears an RGF folder. I updated the DESCRIPTION file of the RGF folder and not the DESCRIPTION file of the R-package folder (it's my first pull request and I'm sorry if this is obvious)

StrikerRUS commented:
@mlampros No problem!

Did you do the following before committing?
>You can apply my changes locally on your compute by either git pull, or just re-clone rgf-1 repo.

mlampros commented:
@StrikerRUS I did the following:

cd /rgf-1                             # this is the fork of rgf
git pull origin master

# I fixed the DESCRIPTION file as you suggested and then,

git add .
git status
git commit -m "fixed DESCRIPTION FILE"
git push -u origin master

StrikerRUS commented:
Strange...
Seems that `gut pull` didn't perform the job. Or something went wrong.

Lets do the following. I'll revert your last commit with duplicated files in wrong folder. After that you'll remove and re-clone the `rgf-1` repository on your machine to make sure you're working with the latest changes. Agree?

mlampros commented:
yes, you can proceed. Notify me once I have to re-fork the rgf.

StrikerRUS commented:
@mlampros 
>... I have to re-fork the rgf repository.

You don't need to **re-fork**, just **re-clone** your repository (`rgf-1`). Since you've proposed this PR, I have write access to the corresponding branch (`master` in our case). It means that I've fixed right in your repository `rgf-1`, and now you should just update your local files. So please `rm -rf rgf-1`, `git clone https://github.com/mlampros/rgf-1.git` and continue working as usual,

mlampros commented:
@StrikerRUS thanks for the info. I've just re-cloned the rgf-1 repository, updated the DESCRIPTION file and pushed the changes.

StrikerRUS commented on R-package/LICENSE.note:
```diff
@@ -0,0 +1,84 @@
+
```
Can we just do something like this:
```
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
#       Regularized Greedy Forest (RGF) 
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 

See file COPYING: https://github.com/RGF-team/rgf/blob/master/RGF/COPYING

...
```
to avoid information duplication?

mlampros replied:
fixed and pushed changes.

StrikerRUS replied:
Thank you for notifying, but you may not to waste your time typing this, I'll see pushed commits anyway. =)

StrikerRUS commented on R-package/NAMESPACE:
```diff
@@ -0,0 +1,11 @@
+
```
Please remove blank line at the top of the file.

mlampros replied:
fixed and pushed changes

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,32 @@
+
+
```
Please remove 2 blank lines at the top of the file.

mlampros replied:
fixed and pushed changes.

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,32 @@
+
+## RGF 1.0.4
+
+* I modified the license from GPL-3 to MIT to go in accordance with the new structure of the *rgf_python* package. The package includes two files : *LICENSE* (for the RGF R package) and *LICENSE.note* (for the *RGF*, *FastRGF* and *rgf_python* packages).
```
Please rewrite this file either in passive voice, or at least with "we" instead of "I"

mlampros replied:
I added "We (RGF-team) ...." for this first case and I pushed the changes.

mlampros replied:
You are probably right. I converted all 'I' to 'We'

StrikerRUS replied:
Thank you!

StrikerRUS commented on R-package/R/package.R:
```diff
@@ -0,0 +1,29 @@
+
```
Blank line.

mlampros replied:
fixed and pushed changes.

StrikerRUS commented on R-package/R/package.R:
```diff
@@ -0,0 +1,29 @@
+
+  if (reticulate::py_available(initialize = TRUE)) {
+
+    if (reticulate::py_module_available("rgf.sklearn")) {
```
Will this function raise understandable error about the need of `rgf` installation? If not, then I think you should write `else` branch and do it there manually.

mlampros replied:
If you don't mind I would rather keep this file intact. Otherwise it might lead to problems when I have to check-test the package on [CRAN](https://cran.r-project.org/)

StrikerRUS replied:
Got it!

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,32 @@
+
+## RGF 1.0.0
+
+
```
`Initial version` maybe?

mlampros replied:
The initial version in all of my R packages begin from 1.0.0 (this again depends on the maintainer of the R package). There is also the possibility to start from 0.1.0. however I follow the 1.0.0 route.

StrikerRUS replied:
Sorry, maybe I was unclear, but my suggestion was to write something **under** the title `1.0.0`, don't leave empty space in changelog.

mlampros replied:
ok, I'll remove the empty spaces.

mlampros commented:
@StrikerRUS,

I included the scipy ( >= 1.0.0) and numpy ( >= 1.14.0) versions because the RGF R package is tested with these versions and works properly. Why this should be a problem? I guess that the newest versions of numpy and scipy are installed when someone installs scikit-learn.

StrikerRUS commented:
It's abnormal to require rather old version of `scikit-learn` and at the same time the newest versions of supporting dependencies.
One can have environment where by any reason old versions of listed packages might be needed. After reading your docs one will have to create another environment with newest versions, which is inconvenient and not required actually. We should do our best to support as many possible situations as possible.

So you should check whether you are using any functions from the newest `numpy`, `scipy`. If not, downgrade required versions to those which are needed for `scikit-learn 0.18`.

StrikerRUS commented on R-package/LICENSE.note:
```diff
@@ -0,0 +1,65 @@
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+Copyright (c) 2016 Baidu, Inc. All Rights Reserved. 2018 RGF-team
```
Please do the same as with `RGF`: just a link, note the whole text.

StrikerRUS commented on R-package/LICENSE.note:
```diff
@@ -0,0 +1,65 @@
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+Copyright (c) 2018 RGF-team and other contributors
```
The same about the link.

StrikerRUS commented on R-package/LICENSE.note:
```diff
@@ -0,0 +1,65 @@
+
+
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Wrapper of machine learning algorithms RGF and FastRGF
```
Either `Python wrapper ...`, or just `rgf_python`.

mlampros commented:
yes, I guess (as you said) excluding the versions from scipy, numpy might be better.

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,26 @@
+I added links to the github repository (master repository, issues).
+
+
+## RGF 1.0.0
```
Oh, seems we have misunderstanding again :-(
I proposed:
Change 
```
## RGF 1.0.0
```
to
```
## RGF 1.0.0

Initial version.
```

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#' @param sl2 a float or NULL. Override L2 regularization parameter l2 for the process of growing the forest. That is, if specified, the weight correction process uses l2 and the forest growing process uses sl2. If NULL, no override takes place and l2 is used throughout training.
+#' @param normalize a boolean. If True, training targets are normalized so that the average becomes zero.
+#' @param min_samples_leaf an integer or a float. Minimum number of training data points in each leaf node. If an integer, then consider \emph{min_samples_leaf} as the minimum number. If a float, then \emph{min_samples_leaf} is a percentage and ceil(min_samples_leaf * n_samples) are the minimum number of samples for each node.
+#' @param n_iter an int or NULL. The number of iterations of coordinate descent to optimize weights. If NULL, 10 is used for loss = "LS" and 5 for loss = "Expo" or "Log".
```
`an int` -> `an integer`

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#' 
+#' the \emph{dump_model} function currently prints information about the fitted model in the console
+#'
+#' @references \emph{https://github.com/fukatani/rgf_python}, \emph{Rie Johnson and Tong Zhang, Learning Nonlinear Functions Using Regularized Greedy Forest}
```
fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#' @param sl2 a float or NULL. Override L2 regularization parameter l2 for the process of growing the forest. That is, if specified, the weight correction process uses l2 and the forest growing process uses sl2. If NULL, no override takes place and l2 is used throughout training.
+#' @param normalize a boolean. If True, training targets are normalized so that the average becomes zero.
+#' @param min_samples_leaf an integer or a float. Minimum number of training data points in each leaf node. If an integer, then consider \emph{min_samples_leaf} as the minimum number. If a float, then \emph{min_samples_leaf} is a percentage and ceil(min_samples_leaf * n_samples) are the minimum number of samples for each node.
+#' @param n_iter an int or NULL. The number of iterations of coordinate descent to optimize weights. If NULL, 10 is used for loss = "LS" and 5 for loss = "Expo" or "Log".
```
`an int` -> `an integer`

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#'
+#' the \emph{predict_proba} function predicts class probabilities for x.
+#'
+#' the \emph{cleanup} function removes tempfiles used by this model. See the issue \emph{https://github.com/fukatani/rgf_python/issues/75}, which explains in which cases the \emph{cleanup} function applies.
```
fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#' 
+#' the \emph{dump_model} function currently prints information about the fitted model in the console
+#'
+#' @references \emph{https://github.com/fukatani/rgf_python}, \emph{Rie Johnson and Tong Zhang, Learning Nonlinear Functions Using Regularized Greedy Forest}
```
fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#'
+#' the \emph{predict} function predicts the regression target for x.
+#'
+#' the \emph{cleanup} function removes tempfiles used by this model. See the issue \emph{https://github.com/fukatani/rgf_python/issues/75}, which explains in which cases the \emph{cleanup} function applies.
```
fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#'
+#' the \emph{score} function returns the coefficient of determination ( R^2 ) for the predictions.
+#'
+#' @references \emph{https://github.com/fukatani/rgf_python}, \emph{Tong Zhang, FastRGF: Multi-core Implementation of Regularized Greedy Forest (https://github.com/baidu/fast_rgf)}
```
fix urls

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+                                     }
+
+
+                                     # initialize RGF_Regressor
```
FastRGF_Regressor

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#'
+#' the \emph{predict_proba} function predicts class probabilities for x.
+#'
+#' the \emph{cleanup} function removes tempfiles used by this model. See the issue \emph{https://github.com/fukatani/rgf_python/issues/75}, which explains in which cases the \emph{cleanup} function applies.
```
fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+#'
+#' the \emph{score} function returns the mean accuracy on the given test data and labels.
+#'
+#' @references \emph{https://github.com/fukatani/rgf_python}, \emph{Tong Zhang, FastRGF: Multi-core Implementation of Regularized Greedy Forest (https://github.com/baidu/fast_rgf)}
```
fix urls

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+                                      }
+
+
+                                      # initialize RGF_Regressor
```
FastRGF_Classifier

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,843 @@
+                                  )
+)
+
+
```
2 blank lines at the end of the file

StrikerRUS commented on R-package/R/zzz.R:
```diff
@@ -0,0 +1,7 @@
+
```
blank line at the top of the file

StrikerRUS commented on R-package/R/zzz.R:
```diff
@@ -0,0 +1,7 @@
+.onAttach <- function(libname, pkgname) {
+  
+  packageStartupMessage("Begninning from version 1.0.3 the 'dgCMatrix_2scipy_sparse' function was renamed to 'TO_scipy_sparse' and now accepts either a 'dgCMatrix' or a 'dgRMatrix' as input. The appropriate format for the 'RGF' package in case of sparse matrices is the 'dgCMatrix' format (scipy.sparse.csc_matrix)")
+}
```
need 1 blank line at the end of the file

mlampros commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,26 @@
+## RGF 1.0.2
+
+I commented the example(s) and test(s) related to the *dgCMatrix_2scipy_sparse* function [ *if (Sys.info()["sysname"] != 'Darwin')* ], because the *scipy-sparse* library on CRAN is not upgraded and the older version includes a bug (*TypeError : could not interpret data type*). This leads to an error on *Macintosh* Operating System ( *reference* : https://github.com/scipy/scipy/issues/5353 ).
+
```
I'm sorry but I'm kind of retro [https://en.wikipedia.org/wiki/Macintosh](https://en.wikipedia.org/wiki/Macintosh)

StrikerRUS replied:
**Macintosh** is a *is a family of personal computers designed*, while **macOS** is *a series of graphical operating systems developed and marketed by Apple Inc.*

You're writing
>This leads to an error on *Macintosh* Operating System

, so it should be **macOS**.

mlampros replied:
Probably here is more clearly stated [https://en.wikipedia.org/wiki/Macintosh#Software](https://en.wikipedia.org/wiki/Macintosh#Software). I don't use Macintosh. I use Linux on a daily basis, however I can't stand calling the Macintosh family (Hardware, Software) as MacOS. If this does not cause any problems to the rgf_python package I would rather prefer that it stays as is.

StrikerRUS replied:
So, your phase
> This leads to an error on Macintosh Operating System

means "Operating System which runs on Macintosh", not the name of OS, right?

StrikerRUS replied:
OK, leave it as is.

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,840 @@
+#'
+#' the \emph{predict} function predicts the regression target for x.
+#'
+#' the \emph{cleanup} function removes tempfiles used by this model. See the issue \emph{https://github.com/fukatani/rgf_python/issues/75}, which explains in which cases the \emph{cleanup} function applies.
```
Fix url

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,840 @@
+                                      }
+
+
+                                      # initialize FastRGF_Regressor
```
FastRGF_Classifier

StrikerRUS commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,840 @@
+                                                                                   as.integer(sparse_min_occurences), calc_prob, as.integer(n_jobs), as.integer(verbose))
+                                    }
+                                  )
+)
```
1 blank line at the end of the file
![image](https://user-images.githubusercontent.com/25141164/43592944-504ab852-967f-11e8-8fb9-1a55a8789c71.png)

StrikerRUS commented on R-package/LICENSE:
```diff
@@ -0,0 +1,2 @@
+YEAR: 2018
+COPYRIGHT HOLDER: Mouselimis Lampros <mouselimislampros@gmail.com>
```
What is the license of the RGF R wrapper?

mlampros replied:
The license is MIT. This appears in the [DESCRIPTION](https://github.com/mlampros/rgf-1/blob/master/R-package/DESCRIPTION) file (License: MIT + file LICENSE)

StrikerRUS replied:
OK, then please put the text of the MIT license here, because after opening `LICENSE` file everyone expects to see the license itself.

mlampros replied:
I'm sorry but I can't do that. CRAN requires from R package maintainers to only mention the YEAR and the COPYRIGHT HOLDER. If I'll add the MIT text inside the LICENSE file then the package will not pass the tests and the package will not be accepted on CRAN. This is actually the reason that the DESCRIPTION file includes *License: MIT + file LICENSE*

StrikerRUS replied:
Are you sure in this? Seems that it's allowed to add test of a license. Please take a look:
https://github.com/dmlc/xgboost/blob/master/R-package/LICENSE

mlampros replied:
This is an Apache license. I'm not familiar with this schema. However in case of the MIT license the [approved schema](https://cran.r-project.org/web/licenses/MIT) is the following : *... This is a template. Complete and ship as file LICENSE the following 2 lines (only) ...*

StrikerRUS commented on R-package/tests/testthat/test_package.R:
```diff
@@ -0,0 +1,354 @@
+
```
Blank line at the top of the file

StrikerRUS commented on R-package/tests/testthat/test_package.R:
```diff
@@ -0,0 +1,354 @@
+    testthat::expect_true( length(pr) == length(y_reg) && sum(validate) == 15 && is.double(tmp_score) && is.double(tmp_score_W) )
+  })
+}
+  
```
Excess blank line at the end of the file

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,566 @@
+
+
+
+The RGF package is a wrapper of the [*Regularized Greedy Forest*](https://github.com/fukatani/rgf_python) *python* package, which also includes a [*Multi-core implementation (FastRGF)*](https://github.com/baidu/fast_rgf). Portability from Python to R was made possible using the [reticulate](https://github.com/rstudio/reticulate) package and the installation requires basic knowledge of Python. Exc ...
```
Fix urls

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,566 @@
+
+|    RGF_Regressor        |   RGF_Classifier       |  FastRGF_Regressor    |  FastRGF_Classifier  |
+| :-----------------:     |  :-----------------:   | :------------------:  | :------------------: |
+|   fit()                 |  fit(()                | fit()                 | fit()                |
```
excess `(` in the RGF_Classifier's column.

Don't you think that using blank cells will increase the readability of this table? I mean, one row for one concrete method.

mlampros replied:
I'm sorry but I don't understand what the 'excess (' issue means. I don't see any problems in the markdown text,

```R

|    RGF_Regressor        |   RGF_Classifier       |  FastRGF_Regressor    |  FastRGF_Classifier  |
| :---------------------: |  :-------------------: | :-------------------: | :------------------: |
|   fit()                 |  fit(()                | fit()                 | fit()                |
|   predict()             |  predict()             | predict()             | predict()            |
|   cleanup()             |  predict_proba()       | cleanup()             | predict_proba()      |
|   get_params()          |  cleanup()             | get_params()          | cleanup()            |
|   score()               |  get_params()          | score()               | get_params()         |
|   feature_importances() |  score()               |                       | score()              |
|   dump_model()          |  feature_importances() |                       |                      |
|                         |  dump_model()          |                       |                      |  

```

StrikerRUS replied:
Here it is:
![image](https://user-images.githubusercontent.com/25141164/43634769-c2a87538-9715-11e8-9f31-489360e1ada3.png)

mlampros replied:
Thanks for this. The red arrow was pretty helpful here.

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,566 @@
+
+|    RGF_Regressor        |   RGF_Classifier       |  FastRGF_Regressor    |  FastRGF_Classifier  |
+| :-----------------:     |  :-----------------:   | :------------------:  | :------------------: |
+|   fit()                 |  fit(()                | fit()                 | fit()                |
```
Also, all classes from this table have `set_params()` method. Does R wrapper supports them too?

mlampros replied:
I understand. The RGF R package supports all available methods of the rgf_python wrapper. Not all methods have to be included in this table in order to work. I explained how this can be done in a [past issue](https://github.com/mlampros/RGF/issues/2#issuecomment-397206793).

I included all methods that I thought they might important / necessary to make both RGF and FastRGF work. The *set_params()* method might be tricky because it requires that a dictionary is passed to the method ( dictionary does not exist as an object in R and someone has to use the reticulate::dict() function ).

Can you include a use case, except for the [test cases](https://github.com/RGF-team/rgf/blob/master/python-package/tests/test_rgf_python.py) that I'm aware of, to understand the importance of setting the parameters after the algorithm (RGF, FastRGF) is initialized?

StrikerRUS replied:
Oh, I see. I didn't know that passing a dictionary is rather complicated task. Forget my comment =)

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,566 @@
+
+  if (is.factor(RESP)) {
+
+    stop(simpleError("this function is meant for regression for classification 
```
`for regression for classification` any separator should be here.

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
````diff
@@ -0,0 +1,566 @@
+
+```
+
+<br>
````
2 excess blank lines at the end of the file

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,565 @@
+
+|    RGF_Regressor        |   RGF_Classifier       |  FastRGF_Regressor    |  FastRGF_Classifier  |
+| :---------------------: |  :-------------------: | :-------------------: | :------------------: |
+|   fit()                 |  fit(()                | fit()                 | fit()                |
```
I can illustrate my previous thoughts about "one row - one method" for better readability:

![image](https://user-images.githubusercontent.com/25141164/43635033-9878656a-9716-11e8-942e-d6fae7d1ba76.png)

StrikerRUS commented:
Thank you very much for your fixes!

According to the notes in each `Rd` file
>% Generated by roxygen2: do not edit by hand 

I suppose that you should regenerate them now and push then here.

Also please share your thoughts about this https://github.com/RGF-team/rgf/pull/208#discussion_r207158210

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,565 @@
+
+
+
+The RGF package is a wrapper of the [*Regularized Greedy Forest*](https://github.com/RGF-team/rgf/tree/master/python-package) *python* package, which also includes a [*Multi-core implementation (FastRGF)*](https://github.com/RGF-team/rgf/tree/master/FastRGF). Portability from Python to R was made possible using the [reticulate](https://github.com/rstudio/reticulate) package and the installation r ...
```
GitHub

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
```diff
@@ -0,0 +1,565 @@
+
+<br>
+
+For the high-dimensional data (can be downloaded from my [Github repository](https://github.com/mlampros/DataSets)) I'll use the *FastRGF_Regressor* rather than the RGF_Regressor (comparison **without doing hyper-parameter tuning**),
```
GitHub

StrikerRUS replied:
The same thing about `my` and `I'll`

StrikerRUS commented on R-package/vignettes/the_RGF_package.Rmd:
````diff
@@ -0,0 +1,565 @@
+
+```{r, eval = F, echo = T}
+
+# download the data from my Github repository (tested on a Linux OS)
````
GitHub

StrikerRUS replied:
The same thing about `my`

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,28 @@
+
+## RGF 1.0.1
+
+We added links to the github repository (master repository, issues).
```
GitHub

StrikerRUS commented on R-package/README.md:
````diff
@@ -0,0 +1,564 @@
+```
+<br>
+
+and to download the latest version from Github use the *install_github* function of the devtools package,
````
GitHub

StrikerRUS commented on R-package/README.md:
````diff
@@ -0,0 +1,564 @@
+```R
+
+
+# download the most recent version of rgf-python from the github repository 
````
GitHub

StrikerRUS commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,28 @@
+## RGF 1.0.4
+
+* We (RGF-team) modified the license from GPL-3 to MIT to go in accordance with the new structure of the *rgf_python* package. The package includes two files : *LICENSE* (for the RGF R package) and *LICENSE.note* (for the *RGF*, *FastRGF* and *rgf_python* packages).
```
I think we don't need this clarification anymore: `We (RGF-team) modified` -> `We modified`

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+<br>
+
+On windows the user can take advantage of the RGF package currently **only** from within the command prompt (console). First, find the full path of the installation location of R (possible if someone right-clicks in the R short-cut (probably on Desktop) and navigates to properties >> shortcut >> target). In my OS, for instance, R is located in **C:\\Program Files\\R\\R-3.4.0\\bin\\x64\\R**. Then, ...
```
Please do something with ` In my OS` and  `in my case` because it's not clear who am I.

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+<br>
+
+#### **Debian/Ubuntu/Fedora**    [ Python 2.7 ]
```
I think that now you can just give a link to the `rgf_python` installation guide and write here only those notes which are related to the **R-package**, like `After a successful rgf-python installation the user should open an R session and give the following *reticulate* command `, `**NOTE : CURRENTLY THE PACKAGE ON WINDOWS CAN BE USED ONLY FROM THE COMMAND LINE (cmd)** ` and others. Generally, it's bad idea to duplicate the information which is already can be found in the neighbor folder; providing a link is enough. Doing this you'll be free from fetching changes from upstream guides.

mlampros replied:
I think that it's good that you commented on this case and I believe that now @fukatani should also participate in this pull request / discussion. 
My opinion to this matter is that the installation instructions of RGF and FastRGF should be more detailed.
I say this because it took me personally many hours to install RGF on different Operating Systems and different python versions and I think that users (like me for instance) should not spend hours or even days in order to at least test the rgf_python or a future version of the R-package.

StrikerRUS replied:
Yeah, of cource!

We can together refine/enhance installation instructions of RGF/FastRGF/rgf_python, but it's out of the scope of this PR.

mlampros replied:
No, I think it's related because we first have to refine / enhance the installation instructions of the RGF / FastRGF / rgf_python and then we should proceed with the update / modification of the installation instructions in the R-package.

StrikerRUS replied:
Unfortunately, we won't mix up different things in one PR making it extremely large.

But reworking the R package guide after enhancing other is good point. So, I suggest you only fix stuff connected with migration (paths/urls), but remained refactoring will be processed in next PR.

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+
+
+#### **Macintosh OSX**            [ Python >= 3.4 ]
```
Can you please elaborate why RGF can't work on macOS with Python 2.7?

mlampros replied:
I already mentioned that the RGF R package can be installed on 

```R
Python (2.7 or >= 3.4)

```
The version on the brackets mentions the Python version that I personally installed / tested the RGF R package.

StrikerRUS replied:
It's not obvious. You should explicitly write about it. Because it's can be interpreted that on Linux RGF works with Python (2.7 or >= 3.4), but on MacOS only with  Python (>= 3.4).

mlampros replied:
yes you are right on this.

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+<br><br>
+
+#### **Windows OS**            [ Python >= 3.4 ]
```
The same question for Windows.

mlampros replied:
same as previous comment

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+<br>
+
+**NOTE : CURRENTLY THE PACKAGE ON WINDOWS CAN BE USED ONLY FROM THE COMMAND LINE (cmd)**
```
Why?

mlampros replied:
The rgf_python package once installed on Windows can be used from an IDE or GUI. However, I wasn't able to make the RGF R package to work from within an IDE (such as Rstudio) or the R GUI.
I tried many things, such as running as administrator or with a different python version, however it was impossible. Probably I should open an issue with a help wanted badge.

StrikerRUS replied:
Got it! I assumed that maybe it's some known reticulate restriction or anything else...

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+[https://github.com/mlampros/RGF/issues](https://github.com/mlampros/RGF/issues)
+
+<br>
+
```
Excess 2 blank lines and `<br>` code at the end of the file

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+Use the following link to report bugs/issues,
+<br><br>
+
+[https://github.com/mlampros/RGF/issues](https://github.com/mlampros/RGF/issues)
```
Update url

StrikerRUS commented on R-package/README.md:
````diff
@@ -0,0 +1,564 @@
+
+```R
+
+devtools::install_github(repo = 'mlampros/RGF')
````
Update command

mlampros replied:
can you please elaborate on which the web link address will be because the R package does not currently exist.

StrikerRUS replied:
Sure:
```
devtools::install_github(repo = 'RGF-team/rgf', subdir = 'R-package')
```

mlampros replied:
thanks.

mlampros commented on R-package/README.md:
```diff
@@ -0,0 +1,564 @@
+
+<br>
+
+**References:**
```
I understand. However, I thought it's a better idea to add the references before the installation instructions. The latter take too much place in the README.md file.

StrikerRUS replied:
When you shrink the guide and leave only related to R instructions, the size will be OK.

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,561 @@
+
+<br>
+
+After the new gcc is installed the user should continue with the installation of *rgf-python*,
```
Many thanks for your fixes!

But can you please re-generate `Rd` files, so we might go further?

In addition, last minor fix: `rgf_python`, not `rgf-python`.

StrikerRUS commented:
@mlampros Thank you very much for your patience and work!

I've managed to integrate your `.travis.yml` config file in our general pipeline (via bash script) and I need to push it. After that you'll have to [**re-clone** the repo](https://github.com/RGF-team/rgf/pull/208#issuecomment-409206804) again. Are you OK with this or should I wait?

mlampros commented:
@StrikerRUS you can proceed, if you or @fukatani do not have anything else to modify / correct. If you don't mind, I'd like that you notify me once I have to re-clone the repo.

StrikerRUS commented:
@mlampros I've pushed Travis script, please re-clone the repo.

Unfortunately, tests fail. Could you please take a look?

mlampros commented:
@StrikerRUS I pulled the changes.
My only guess is that the package tests probably fail because the R package is tested with python 3 on linux. Would you mind change the python version to 2.7?

StrikerRUS commented:
@mlampros Unfortunately, the same fails with Python 2.7. Any other ideas?

mlampros commented:
@StrikerRUS the [scipy tests](https://github.com/mlampros/rgf-1/blob/master/R-package/tests/testthat/test_package.R#L248-L313) (4 out of 11) run successfully,

```R
 ══ testthat results  ═══════════════════════════════════════════════════════════
  OK: 4 SKIPPED: 0 FAILED: 7
  1. Error: the methods of the 'RGF_Regressor' class return the correct output (@test_package.R#90) 
  2. Error: the methods of the 'RGF_Classifier' class return the correct output (binary classification) (@test_package.R#120) 
  3. Error: the methods of the 'RGF_Classifier' class return the correct output (multiclass classification) (@test_package.R#147) 
  4. Error: the methods of the 'FastRGF_Regressor' class return the correct output (@test_package.R#179) 
  5. Error: the methods of the 'FastRGF_Classifier' class return the correct output (binary classification) (@test_package.R#211) 
  6. Error: the methods of the 'FastRGF_Classifier' class return the correct output (multiclass classification) (@test_package.R#239) 
  7. Error: the RGF_Regressor works with sparse (scipy) matrices (@test_package.R#346) 

```

which means that the reticulate package finds the correct python configuration. Therefore tests fail for the *rgf.sklearn* or *rgf.utils* modules ( [package.R](https://github.com/mlampros/rgf-1/blob/master/R-package/R/package.R) ). I guess the R-package can not find the rgf-python package.
The fact is that you use [miniconda](https://github.com/mlampros/rgf-1/blob/master/.travis.yml) in the *before_install* step [whereas I don't ](https://github.com/mlampros/RGF/blob/master/.travis.yml) and the installation of the R package dependencies, in my case, takes place in the [*install* step](https://github.com/mlampros/RGF/blob/master/.travis.yml)

StrikerRUS commented:
>I guess the R-package can not find the rgf-python package.

How we can check this assumption?

>The fact is that you use miniconda in the before_install step whereas I don't and the installation of the R package dependencies, in my case, takes place in the install step

How it can interfere the RGF R-package installation?

StrikerRUS commented:
>the failed tests is the answer.

Nope. Tests can fail by any other reasons.

I figured it out that `reticulate` really cannot find `rgf_python`:

```
Rscript -e 'reticulate::py_discover_config(required_module="rgf_python")'
```

returns

```
python:         /home/travis/miniconda/envs/test-environment/bin/python
libpython:      /home/travis/miniconda/envs/test-environment/lib/libpython2.7.so
pythonhome:     /home/travis/miniconda/envs/test-environment:/home/travis/miniconda/envs/test-environment
version:        2.7.15 |Anaconda, Inc.| (default, May  1 2018, 23:32:55)  [GCC 7.2.0]
numpy:          /home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/numpy
numpy_version:  1.15.0
rgf_python:     [NOT FOUND]
```

StrikerRUS commented:
> I'm just telling what it worked for me when I built the RGF R package on travis.

You can find that all these steps are placed now in `.R.travis.sh`

StrikerRUS commented:
Also, it's very strange why tests fail instead of be marked as `SKIPPED`: https://github.com/mlampros/rgf-1/blob/a2a3798a4dbac1cd7924d47aec8fea60e1e65cb8/R-package/tests/testthat/test_package.R#L76

StrikerRUS commented:
@mlampros Is there any way to make installation process more verbose and see full error traceback?

mlampros commented:
@StrikerRUS I'm sorry for the late reply. The fact is that I'm not familiar with the Anaconda / Miniconda environment. I do know only the basics. 
In the reticulate repository there appears to be [related past issues](https://github.com/rstudio/reticulate/issues?utf8=%E2%9C%93&q=conda+).
I don't see a verbose parameter in the [R CMD INSTALL](https://support.rstudio.com/hc/en-us/articles/200486518-Customizing-Package-Build-Options) command, but it seems that Rscript has a verbose parameter,

```R
Rscript --help


--options accepted are
  --help              Print usage and exit
  --version           Print version and exit
  --verbose           Print information on progress
  --default-packages=list
                      Where 'list' is a comma-separated set
                        of package names, or 'NULL'
or options to R, in addition to --slave --no-restore, such as
  --save              Do save workspace at the end of the session
  --no-environ        Don't read the site and user environment files
  --no-site-file      Don't read the site-wide Rprofile
  --no-init-file      Don't read the user R profile
  --restore           Do restore previously saved objects at startup
  --vanilla           Combine --no-save, --no-restore, --no-site-file
                        --no-init-file and --no-environ

```

StrikerRUS commented:
Thank you for the link, I'll try to find something useful there. It's weird that reticulate finds rgf, but tests fail. I suppose that the reason for this is in something else. Can you please try to get the same version of R on your machine and try to find rgf `Rscript -e 'reticulate::py_discover_config(required_module="rgf")'`?

mlampros commented:
The output of 

```R
Rscript -e 'reticulate::py_discover_config(required_module="rgf")'

```
on my Linux OS is,

```R
python:         /usr/bin/python
libpython:      /usr/lib/python2.7/config-x86_64-linux-gnu/libpython2.7.so
pythonhome:     /usr:/usr
version:        2.7.12 (default, Dec  4 2017, 14:50:18)  [GCC 5.4.0 20160609]
numpy:          /usr/local/lib/python2.7/dist-packages/numpy
numpy_version:  1.14.3
rgf:            /usr/local/lib/python2.7/dist-packages/rgf

python versions found: 
 /usr/bin/python
 /usr/bin/python3
```

And the following for *rgf.sklearn* and *rgf.utils*,

```R
Rscript -e 'reticulate::py_module_available("rgf.sklearn")'

```

```R
[1] TRUE

```


```R
Rscript -e 'reticulate::py_module_available("rgf.utils")'

```


```R
[1] TRUE

```

StrikerRUS commented:
Finally, I've managed to set up R environment for checks. R is built from sources and any version can be chosen via `R_VER` variable (cache cleaning is needed beforehand).

The only thing remained to fix is curl certificates.

Also, CRAN check reports about insufficient package version. Fix is needed.

```
Insufficient package version (submitted: 1.0.4, existing: 1.0.4)
```

StrikerRUS commented:
OK, I don't know how to fix certificates issue 😢 

@mlampros Please, fix this problem
```
Found the following (possibly) invalid URLs:
  URL: http://cran.r-project.org/package=RGF
    From: README.md
    CRAN URL not in canonical form
```

and comment out these lines in `.R.travis.sh` for successful build
```
elif grep -q -R "NOTE" "$LOG_FILE_NAME"; then
    echo "NOTES have been found in the build log!"
    exit -1
```

StrikerRUS commented:
@fukatani Please take a look at this PR.

fukatani commented on R-package/DESCRIPTION:
```diff
@@ -0,0 +1,25 @@
+Title: Regularized Greedy Forest
+Version: 1.0.5
+Date: 2018-08-09
+Authors@R: c( person("Lampros", "Mouselimis", email = "mouselimislampros@gmail.com", role = c("aut", "cre")), person("Ryosuke", "Fukatani", role = "cph", comment = "Author of the python wrapper of the 'Regularized Greedy Forest' machine learning algorithm"), person("Tong", "Zhang", role = "cph", comment = "Author of the 'Regularized Greedy Forest' and of the Multi-core implementation of Regulariz ...
```
@StrikerRUS Your name should be added? You are also "Author of the python wrapper of the 'Regularized Greedy Forest' machine learning algorithm".

StrikerRUS replied:
@fukatani Thank you! It'll be an honor for me.

fukatani commented on R-package/NEWS.md:
```diff
@@ -0,0 +1,33 @@
+
+* We modified the license from GPL-3 to MIT to go in accordance with the new structure of the *rgf_python* package. The package includes two files : *LICENSE* (for the RGF R package) and *LICENSE.note* (for the *RGF*, *FastRGF* and *rgf_python* packages).
+* We added the following new features of RGF estimators : *feature_importances_* and *dump_model()*
+* We modified the README.md file and especially the installation instructions for all operating systems (Linux, Macintosh, Windows)
```
MacOS is better.

fukatani commented on R-package/R/zzz.R:
```diff
@@ -0,0 +1,6 @@
+
+.onAttach <- function(libname, pkgname) {
+  
+  packageStartupMessage("Begninning from version 1.0.3 the 'dgCMatrix_2scipy_sparse' function was renamed to 'TO_scipy_sparse' and now accepts either a 'dgCMatrix' or a 'dgRMatrix' as input. The appropriate format for the 'RGF' package in case of sparse matrices is the 'dgCMatrix' format (scipy.sparse.csc_matrix)")
```
Typo (Beginning)

fukatani commented on R-package/README.md:
```diff
@@ -0,0 +1,561 @@
+
+
+
+#### **Macintosh OSX**            [ installed / tested on Python >= 3.4 ]
```
Mac OS X is formal name.

fukatani commented on R-package/tests/testthat/test_package.R:
```diff
@@ -0,0 +1,352 @@
+  
+    testthat::expect_true( length(pr) == length(y_reg) && sum(validate) == 15 && is.double(tmp_score) && is.double(tmp_score_W) )
+  })
+}
```
We need irregular arguments test. ex. n_estimators < 0.
It is not necessary to examine all the arguments, but it is better for one argument per estimator class to have such a test.

mlampros replied:
You are right. However, since the R-package ports the Python-package I don't think it's wise to repeat the [same tests](https://github.com/RGF-team/rgf/blob/master/python-package/tests/test_rgf_python.py).

fukatani replied:
OK, you don't have to test rgf_python here.
But what happen in R layer if rgf_python raised exception?

`reticulate` defines behavior in such case?
If so, we don't have to test `reticulate` function.

mlampros replied:
I think an example can solve the issue here,

```R
library(RGF)
set.seed(1)
x = matrix(runif(1000), nrow = 100, ncol = 10)
y = runif(100)
RGF_regr = RGF_Regressor$new(max_leaf = -1)
RGF_regr$fit(x, y)

```
<br>
This is the error that I receive using the Rstudio IDE,

```R
 Show Traceback
 
 Rerun with Debug
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  ValueError: max_leaf must be greater than 0 but was -1. 

```
<br>
Concerning the tests of the *reticulate* package, if you don't mind I would like to keep the tests intact, because I have also to submit the package to CRAN before I update the version for the RGF-team repository and this might lead to issues.

StrikerRUS commented:
All problems with certificates have been solved. arxiv.org certificate has been added to the rest certificates, link to r-pkg.org has been replaced with link to the package official page at CRAN, link to the MinGW-w64 title page has been replaced with http protocol, link to MinGW-w64 download page has been replaced with the direct link to SourceForge.

@mlampros Please fix the WARNING which has been introduced in https://github.com/RGF-team/rgf/pull/208/commits/44703260409275b8db75caddb6acba8e1459de3c
```
* checking for missing documentation entries ... WARNING
Undocumented code objects:
  ‘FastRGF_Classifier’ ‘FastRGF_Regressor’ ‘mat_2scipy_sparse’
  ‘RGF_Classifier’ ‘RGF_cleanup_temp_files’ ‘RGF_Regressor’
  ‘TO_scipy_sparse’
All user-level objects in a package should have documentation entries.
See chapter ‘Writing R documentation files’ in the ‘Writing R
Extensions’ manual.
```

fukatani commented on R-package/R/utils.R:
```diff
@@ -0,0 +1,840 @@
+
+  else {
+
+    stop("the function can take either a 'sparse_row_matrix' or a 'sparse_column_matrix' for the 'format' parameter as input", call. = F)
```
Is this logic tested?

mlampros replied:
yes, it is [tested](https://github.com/mlampros/rgf-1/blob/master/R-package/tests/testthat/test_package.R#L251-L256)

fukatani commented on R-package/.R.travis.sh:
```diff
@@ -0,0 +1,60 @@
+    exit -1
+#elif grep -q -R "NOTE" "$LOG_FILE_NAME"; then
+#    echo "NOTES have been found in the build log!"
+#    exit -1
```
If unnecessary, l.55-l.57 should be deleted.

mlampros replied:
I guess @StrikerRUS should decide if this code chunk should be removed

StrikerRUS commented on R-package/README.md:
```diff
@@ -0,0 +1,561 @@
+
+<br><br>
+
+#### **Windows OS**            [ installed / tested on Python >= 3.4 ]
```
Yeah, lets do this in next PR. not this one.

fukatani commented on R-package/.R.travis.sh:
```diff
@@ -0,0 +1,60 @@
+#    exit -1
+fi
+
+Rscript -e 'covr::codecov(quiet = FALSE)'
```
This line can be related to codecov behavior.

StrikerRUS replied:
This line has been here since the first commit in the PR. Before today it silently printed information to the Travis log and updated the badge, but today decided to come here in comments and post the report.

PS. it seems that registration at codecov.io isn't required and done automatically after using the package.

StrikerRUS replied:
I want to rollback to the previous behavior: report is printed to the Travis log and badge updates without these useless and annoying comments.

StrikerRUS replied:
I've found the way to easily disable comments, but this PR should be merged first.

mlampros commented:
@StrikerRUS the main reason that I use the 3.2.3 version is because I don't want to force users to use the latest version (currently 3.5.1) but on the other hand to avoid using a too old version, as new features and bugs are introduced / fixed from one version to another. You can have a look to the  [news sections](https://cran.r-project.org/doc/manuals/r-devel/NEWS.html).

StrikerRUS commented:
@mlampros Sorry, didn't get it. I'm speaking about earlier versions. On what concrete features/bug fixes of 3.2.3 version RGF is relied?

OK, I'll paraphrase. Imagine, I'm using R 3.2.0 and you are forcing me to upgrade it to 3.2.3. Why? The fact is that RGF can work with my version of R. You cannot set min version out of nothing.

mlampros commented:
@StrikerRUS, every time I have to test-check any package that I maintain I've to test it with the latest R-devel version. I guess the 3.2.3 version is for me, as I said, not the newest but also not the oldest one. However, the RGF package imports the following packages,

* reticulate (requires R >= 3.0)
* R6 (requires R >= 3.0)
* Matrix (requires R >= 3.2.0)

You are probably right on that I've have to update the minimum version to 3.2.0, however I didn't like the tone of your comment.

StrikerRUS commented:
@mlampros 
>however I didn't like the tone of your comment.

Sorry, I didn't want to hurt you! Probably it's due to my non-perfect English :-)

Speaking about this,
>Matrix (requires R >= 3.2.0)

it's true only for the two latest (1.2-13, 1.2-14) versions of the Matrix package. Is there any reason to force users use them? Because 1.2-12 version, which was published in November 2017 (I think rather fresh), requires R >= 3.0.1: https://github.com/cran/Matrix/blob/24289cf538a36e90ad4bf8041f51d3397559fd9e/DESCRIPTION

fukatani commented:
Thanks to @mlampros and @StrikerRUS , this PR It seems that it has become quality that can be merged in almost.

R version is still as it is, and expansion of support can be discussed in another PR.

StrikerRUS commented:
@fukatani Yeah! Absolutely agree! My browser even freezes when opens this PR discussion - so many comments 😄 

I'll create separate issues for some comments, which left undiscussed here.
